app/controllers/devise_sessions_controller.rb

```diff
@@ -3,54 +3,79 @@ class DeviseSessionsController < Devise::SessionsController
   include CookiesHelper
 
   before_action :check_login_state, only: %i[
-    create
     phone_code
-    phone_code_send
-    phone_verify
-    phone_resend
-  ]
-
-  before_action :check_password_ok, only: %i[
-    phone_code
-    phone_code_send
     phone_verify
     phone_resend
+    phone_resend_code
   ]
 
   def create
-    render :new and return unless params.dig(:user, :password)
+    payload = get_payload
+
+    render :new and return unless params.dig(:user, :email) || params.dig(:user, :password)
+
+    @email = params.dig(:user, :email)
+    catch(:warden) do
+      self.resource = warden.authenticate(auth_options)
+    end
 
-    self.resource = warden.authenticate(auth_options)
     if resource
-      login_state.update!(password_ok: true)
+      user = User.find_by(email: @email).id
+      @login_state = LoginState.create!(
+        created_at: Time.zone.now,
+        user_id: user,
+        redirect_path: after_login_path(payload, user),
+        jwt_id: session[:jwt_id],
+      )
+
+      @login_state_id = login_state.id
+      session[:login_state_id] = login_state.id
+
       if request.env["warden.mfa.required"]
         MultiFactorAuth.generate_and_send_code(resource)
         redirect_to user_session_phone_code_path
       else
         do_sign_in
       end
     else
-      @password_error_message =
-        case User.find_by(email: params.dig(:user, :email))&.unauthenticated_message
-        when :last_attempt
-          I18n.t("devise.failure.last_attempt")
-        when :locked
-          I18n.t("devise.failure.locked")
-        else
-          I18n.t("devise.sessions.new.fields.password.errors.incorrect")
-        end
-      render :new
-    end
-  end
+      @resource_error_messages = {}
 
-  def phone_code; end
+      if params.dig(:user, :password).blank?
+        @resource_error_messages[:password] = [I18n.t("activerecord.errors.models.user.attributes.password.blank")]
+      end
 
-  def phone_code_send
-    MultiFactorAuth.generate_and_send_code(login_state.user)
+      user = User.find_by(email: @email)
+      user_exists = user.present?
+
+      if user_exists && !user.active_for_authentication?
+        @resource_error_messages[:email] = [I18n.t("devise.failure.unconfirmed")]
+      end
+
+      if user_exists && !user.active_for_authentication?
```

Comment:
We've now got this branch twice

Reply:
Thanks, I managed to get myself all confused with these errors!